server/controllers/v1/user.ts

```diff
@@ -9,13 +11,117 @@ export enum TOKEN_TYPE {
 
 export default class UserController {
   protected userRepo: typeof User;
+  secret: string;
   protected emailService: EmailService;
 
   constructor(model: typeof User, emailService: EmailService) {
     this.userRepo = model;
+    this.secret = process.env.SECRET ? process.env.SECRET : "secret";
     this.emailService = emailService;
   }
 
+  /**
+   * Update the last login time of user to the current date.
+   */
+  private updateLastLogin = async (user: User): Promise<void> => {
+    user.lastLogin = new Date();
+    await user.save();
+  };
+
+  /**
+   * Sign in the user with the given credentials if the user exists.
+   * Generate auth token and update last login date.
+   * @returns status code and data payload
+   */
+  async signIn(
+    userName: string,
+    password: string
+  ): Promise<{ status: number; data: any }> {
+    try {
+      const oldUser = await this.userRepo.findOne({
+        where: {
+          userName: userName,
+        },
+      });
+      if (!oldUser) {
+        return { status: 404, data: { message: "User doesn't exist" } };
+      }
+
+      const isPasswordCorrect = await argon2.verify(
+        oldUser.password,
+        password,
+        {
+          type: argon2.argon2id,
+        }
+      );
+      if (!isPasswordCorrect) {
+        return { status: 400, data: { message: "Invalid credentials" } };
+      }
+
+      const token = jwt.sign(
+        { username: oldUser.userName, id: oldUser.id },
+        this.secret,
+        { expiresIn: "1h" }
+      );
+
+      await this.updateLastLogin(oldUser);
+
+      return { status: 200, data: { result: oldUser, token: token } };
```

Comment:
Token should not be sent as part of the JSON response - cookie only.

as an implementation reference: https://cheatcode.co/tutorials/how-to-implement-secure-httponly-cookies-in-node-js-with-express

Reply:
The UserController has no access to `res` so how should I go about calling `res.cookie` without passing it back to routes?

Reply:
Return [token, data]

Reply:
Or, better yet generate the token in the routes file since the token is HTTP specific.